src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.crypto.provider;
+
+import javax.crypto.KDFSpi;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.HKDFParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.KDFParameters;
+import java.security.NoSuchAlgorithmException;
+import java.security.ProviderException;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.InvalidParameterSpecException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * KDF implementation for the HKDF function.
+ * <p>
+ * This class implements the HKDF-Extract and HKDF-Expand functions from RFC
+ * 5869.  This implementation provides the complete Extract-then-Expand HKDF
+ * function as well as Extract-only and Expand-only variants.
+ */
+abstract class HkdfKeyDerivation extends KDFSpi {
+
+    protected final int hmacLen;
+    protected final String hmacAlgName;
+
+    /**
+     * The sole constructor.
+     *
+     * @param kdfParameters
+     *     the initialization parameters (may be {@code null})
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the initialization parameters are inappropriate for this
+     *     {@code KDFSpi}
+     */
+    HkdfKeyDerivation(String hmacAlgName, int hmacLen,
+                      KDFParameters kdfParameters)
+        throws InvalidAlgorithmParameterException {
+        super(kdfParameters);
+        if (kdfParameters != null) {
+            throw new InvalidAlgorithmParameterException(
+                "RFC 5869 has no parameters for its KDF algorithms");
+        }
+        this.hmacAlgName = hmacAlgName;
+        this.hmacLen = hmacLen;
+    }
+
+    /**
+     * Derive a key, returned as a {@code SecretKey}.
+     *
+     * @return a derived {@code SecretKey} object of the specified algorithm
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code KDFParameterSpec} is
```

Comment:
typo: KDFParaneterSpec should be derivationParameterSpec.
BTW, I assume the javadoc is copied from KDFSpi class. I'd suggest to shorten it and match it to the actual implementation instead of the general description as in KDFSpi.
For the last sentence, do you have an example of how the `derivationParameterSpec` for HKDFParameterSpec can specify a type of output that is not a key?

Reply:
Yes, resolved.